### PR TITLE
Add 'regression' tests to SYCL/InvokeSimd

### PR DESCRIPTION
**Pull request by fveselov**
'Regression' category tests

### CODE REVIEW

File: SYCL/InvokeSimd/regression/call_vadd_1d_loop_naive.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+ * 3). invoke_simd functionality is integrated into this test case
+ *     by accumulating wi_ids into simds instead of vector elements.
+ *     There is thus an extra looping step that must occur in order to
+ *     perform the vector additions: each simd of wi_ids must be
+ *     looped through and each wi_id extracted. Then, each wi_id
+ *     is used to compute an absolute offset into the underlying vectors
+ *     A, B, and C. That is, each ESIMD function call actually performs
+ *     VL vector additions.
+ *
+ * This tests is the same as call_vadd_1d/loop.cpp, but passes offset as
```
Comment (sarnex) [inline SYCL/InvokeSimd/regression/call_vadd_1d_loop_naive.cpp]: Typo, should be 1d_loop.cpp not 1d/loop.cpp I think

File: SYCL/InvokeSimd/regression/call_vadd_1d_loop_naive.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+__attribute__((always_inline)) esimd::simd<float, VL>
+ESIMD_CALLEE(float *A, float *B, float *C, int i) SYCL_ESIMD_FUNCTION {
+  esimd::simd<float, VL> a, b;
+  a.copy_from(A + i);
+  b.copy_from(B + i);
+  esimd::simd<float, VL> c = a + b;
+  c.copy_to(C + i);
+
+  return esimd::simd<float, VL>(); // Need to return something; return a dummy.
```
Comment (sarnex) [inline SYCL/InvokeSimd/regression/call_vadd_1d_loop_naive.cpp]: Do we still need to return a dummy or does void work?

File: SYCL/InvokeSimd/regression/debug_symbols.cpp [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+//
+// Check that full compilation works:
+// RUN: %clangxx -fsycl -fno-sycl-device-code-split-esimd -Xclang -fsycl-allow-func-ptr -g %s -o %t.out
+// RUN: env IGC_VCSaveStackCallLinkage=1 IGC_VCDirectCallsOnly=1 %GPU_RUN_PLACEHOLDER %t.out
+//
+// VISALTO enable run
+// RUN: env IGC_VISALTO=63 IGC_VCSaveStackCallLinkage=1 IGC_VCDirectCallsOnly=1 %GPU_RUN_PLACEHOLDER %t.out
+
+/* Tests invoke_simd support in the compiler/headers
+ * Test checks support of debug simbols.
```
Comment (sarnex) [inline SYCL/InvokeSimd/regression/debug_symbols.cpp]: I might have missed it but I don't see us doing any testing for debug symbols here. If not, can we just update the comment to mention we're testing compiling with -g works?

File: SYCL/InvokeSimd/regression/matrix_add.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  // real reason we are using explicit ND range kernels (i.e., nd_range,
+  // nd_item) in this program is because we need to extract the subgroup from
+  // nd_item in order to pass it to invoke_simd(). A local range is made and so
+  // workitems in the global execution range are grouped into workgroups,
+  // however, we do not use any workgroup-specific functionality for the
+  // computation.
+  // Visualize each workgroup as a B x B sub-matrix.
+  constexpr size_t B = 4; // The matrix is divided into 4 x 4 = 16 workgroups.
+  constexpr unsigned GroupSize = B * B; // 64 * 64 = 4,096 workitems.
+  // constexpr unsigned GroupSize = 4 * VL; // VL is the subgroup size, 16; each
```
Comment (sarnex) [inline SYCL/InvokeSimd/regression/matrix_add.cpp]: Should we remove the commented code here?

File: SYCL/InvokeSimd/regression/matrix_multiply_USM.cpp [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+ * simd lanes with multiple row indices and column indices respectively, i.e.,
+ * simd row_indices = [row0, row1, row2, ..., rowN] and simd column_indices =
+ * [col0, col1, col2, ..., colN]. We use these simd lanes as parallel arrays and
+ * compute a dot product for each (row, col) pair. This means that each
+ * invocation of the ESIMD_CALLEE is responsible for computing VL separate dot
+ * products, each of which is packed back into a simd lane which is then
+ * returned to the caller in the SPMD context. Each dot product is returned
+ * individually to the caller in the SPMD context.
+ *
+ * 4/12/2022: The original test case used buffers/accessors and passed a
```
Comment (sarnex) [inline SYCL/InvokeSimd/regression/matrix_multiply_USM.cpp]: Maybe we can remove this comment if it's not relevant anymore

File: SYCL/InvokeSimd/regression/matrix_multiply_USM.cpp [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+  // invoke_simd(), which requires a subgroup argument. A local (execution)
+  // range is defined and consequently workitems in the global execution range
+  // are grouped into workgroups, however, we do not use any this grouping or
+  // any workgroup-specific functionality for this computation (i.e., workgroup
+  // local memory, work-item communication or synchronization within a
+  // workgroup, etc.).
+  // Visualize each workgroup as a B x B sub-matrix.
+  constexpr size_t B = 4; // The matrix is divided into 4 x 4 = 16 workgroups.
+  constexpr unsigned GroupSize = B * B; // workitems in each workgroup.
+  // constexpr unsigned GroupSize = 4 * VL; // VL is the subgroup size, 16; each
```
Comment (sarnex) [inline SYCL/InvokeSimd/regression/matrix_multiply_USM.cpp]: Commented line again

File: SYCL/InvokeSimd/regression/matrix_multiply_accessor_get_pointer.cpp [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+  // invoke_simd(), which requires a subgroup argument. A local (execution)
+  // range is defined and consequently workitems in the global execution range
+  // are grouped into workgroups, however, we do not use any this grouping or
+  // any workgroup-specific functionality for this computation (i.e., workgroup
+  // local memory, work-item communication or synchronization within a
+  // workgroup, etc.).
+  // Visualize each workgroup as a B x B sub-matrix.
+  constexpr size_t B = 4; // The matrix is divided into 4 x 4 = 16 workgroups.
+  constexpr unsigned GroupSize = B * B; // workitems in each workgroup.
+  // constexpr unsigned GroupSize = 4 * VL; // VL is the subgroup size, 16; each
```
Comment (sarnex) [inline SYCL/InvokeSimd/regression/matrix_multiply_accessor_get_pointer.cpp]: Same

File: SYCL/InvokeSimd/regression/tiled_matrix_multiplication.cpp [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+              int m = ndi.get_global_id(0); // Global row index
+              int n = ndi.get_global_id(1); // Global column index
+
+              // Index in the local index space:
+              int i = ndi.get_local_id(1); // Local index
+              int sum = 0;
+
+              if constexpr (use_invoke_simd) {
+                for (int kk = 0; kk < M; kk += tile_size) {
+                  // Load the matrix tile from matrix A, and sychronize
```
Comment (sarnex) [inline SYCL/InvokeSimd/regression/tiled_matrix_multiplication.cpp]: typo in sychronize

File: SYCL/InvokeSimd/regression/tiled_matrix_multiplication.cpp [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+                  // like the only aspect that is parallelizable ACROSS the
+                  // work-items in a work-group. Ideally, in a matrix multiply,
+                  // we want to parallelize the vector dot prodct computation
+                  // carried out by a single work-item, but this is the wrong
+                  // level of granularity for invoke_simd; this is
+                  // intra-work-item granularity, whereas invoke_simd is
+                  // intra-work-group (inter-work-item) granularity.
+                  tileA[i] = invoke_simd(ndi.get_sub_group(), SIMD_CALLEE,
+                                         acc_a[m][kk + i]);
+                  // NOTE: Because invoke_simd implicitly sychronizes the
```
Comment (sarnex) [inline SYCL/InvokeSimd/regression/tiled_matrix_multiplication.cpp]: typo again

File: SYCL/InvokeSimd/regression/tiled_matrix_multiplication.cpp [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+                  // We implicitly parallelize the loading of the tile; each
+                  // work-item i in the work-group loads its corresponding tile
+                  // element from matrix A (acc_a) in global memory. This seems
+                  // like the only place where it makes sense and is easily
+                  // implementable to insert invoke_simd. Remember, invoke_simd
+                  // implicitly parallelizes some aspect of the workload across
+                  // the work-items belonging to the same workgroup. This seems
+                  // like the only aspect that is parallelizable ACROSS the
+                  // work-items in a work-group. Ideally, in a matrix multiply,
+                  // we want to parallelize the vector dot prodct computation
```
Comment (sarnex) [inline SYCL/InvokeSimd/regression/tiled_matrix_multiplication.cpp]: typo: product

**Review by sarnex — APPROVED** [review]

File: SYCL/InvokeSimd/regression/debug_symbols.cpp (hunk 3 above)
Comment (fveselov) [reply SYCL/InvokeSimd/regression/debug_symbols.cpp]: Debug symbols option (**-g**) is passed to clang:
`// RUN: %clangxx -fsycl -fno-sycl-device-code-split-esimd -Xclang -fsycl-allow-func-ptr -g %s -o %t.out`

Comment (sarnex) [reply SYCL/InvokeSimd/regression/debug_symbols.cpp]: Ah sorry, I read the comment as saying it was testing the generated binary to make sure debug info was preserved

File: SYCL/InvokeSimd/regression/call_vadd_1d_loop_naive.cpp (hunk 2 above)
Comment (fveselov) [reply SYCL/InvokeSimd/regression/call_vadd_1d_loop_naive.cpp]: Thanks for pointing out, changed to void return.

**Comment by sarnex** [comment]
LGTM, we can merge once CI passes :)

**Review by v-klochkov — COMMENTED** [review]
I have looked through this huge PR. It looks good to me. Thanks to @sarnex for being main reviewer of this PR.

Maybe just one question that can be fixed easily.
The folder names used in InvokeSimd folder use 3 different styles:
- SomeFolder // GOOD // e.g. 'InvokeSimd'
- something   // Ok // e.g. 'regression'. IMO 'Regression would be more consistent, but Ok.
- SOME_FOLDER // not Ok, IMO // e.g. 'IMPL_SUBGROUP'. Can it be renamed to 'ImplicitSubgroup', please?

File: SYCL/InvokeSimd/regression/call_vadd_1d_spill.cpp [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+// RUN: env IGC_VCSaveStackCallLinkage=1 IGC_VCDirectCallsOnly=1 %GPU_RUN_PLACEHOLDER %t.out
+//
+// VISALTO enable run
+// RUN: env IGC_VISALTO=63 IGC_VCSaveStackCallLinkage=1 IGC_VCDirectCallsOnly=1 %GPU_RUN_PLACEHOLDER %t.out
+
+// Tests invoke_simd support in the compiler/headers
+/* This program is basically an extension of the standard vector addition
+ * program (i.e., call_vadd_1d). The only difference is that instead of adding 2
+ * vectors (i.e., va + vb), we add 14 vectors. However, in this implemenation,
+ * the extra vectors beyond va and vb are zero vectors, so adding them has no
```
Comment (v-klochkov) [inline SYCL/InvokeSimd/regression/call_vadd_1d_spill.cpp]: Would it be difficult to make these extra vectors holding some meaningful non-zero values?
My concern here is that even if part of code or params/operands is lost or corrupted, then the test will not catch that.

Comment (fveselov) [reply SYCL/InvokeSimd/regression/call_vadd_1d_spill.cpp]: There are actually 2 cases in the test: one with zero extra vectors and one with non-zero. I fixed comment to represent real test configuration. Thanks for pointing out.

**Comment by fveselov** [comment]
> I have looked through this huge PR. It looks good to me. Thanks to @sarnex for being main reviewer of this PR.
> 
> Maybe just one question that can be fixed easily. The folder names used in InvokeSimd folder use 3 different styles:
> 
> * SomeFolder // GOOD // e.g. 'InvokeSimd'
> * something   // Ok // e.g. 'regression'. IMO 'Regression would be more consistent, but Ok.
> * SOME_FOLDER // not Ok, IMO // e.g. 'IMPL_SUBGROUP'. Can it be renamed to 'ImplicitSubgroup', please?

Agree, renamed both 'regression' and 'IMPL_SUBGROUP' directories.

I also created one more PR (https://github.com/intel/llvm-test-suite/pull/1562) doing a similar renames for the "Feature" and "Spec" categories of InvokeSimd.

**Comment by fveselov** [comment]
Windows and Linux CIs failures are unrelated to changes:
> [2023-02-01T10:35:43.252Z] Failed Tests (3):
> [2023-02-01T10:35:43.252Z]   SYCL :: ESIMD/api/svm_gather_scatter_scalar_off.cpp
> [2023-02-01T10:35:43.252Z]   SYCL :: ESIMD/dword_atomic_cmpxchg_scalar_off.cpp
> [2023-02-01T10:35:43.252Z]   SYCL :: ESIMD/dword_atomic_smoke_scalar_off.cpp

**Review by v-klochkov — APPROVED** [review]

**Comment by v-klochkov** [comment]
> Windows and Linux CIs failures are unrelated to changes:
> 
> > [2023-02-01T10:35:43.252Z] Failed Tests (3):
> > [2023-02-01T10:35:43.252Z]   SYCL :: ESIMD/api/svm_gather_scatter_scalar_off.cpp
> > [2023-02-01T10:35:43.252Z]   SYCL :: ESIMD/dword_atomic_cmpxchg_scalar_off.cpp
> > [2023-02-01T10:35:43.252Z]   SYCL :: ESIMD/dword_atomic_smoke_scalar_off.cpp

@fineg74 - please take a look at these fails, they must be caused by one of your recent changes in these tests.
Well, it also possible that the fails showed up because the older/yesterday version of compiler was used to run this llvm-test-suite changes.

**Comment by fineg74** [comment]
> 

@v-klochkov It seems the compiler change for scalar offset didn't go into latest nightly build for some reason that is used to run tests. The latest commit that went into nightly build was 8:23 AM commit while scalar offset change was merged at 15:35
Will follow up tomorrow if the problem persists

**Pull request merged**